otter/controller.py

```diff
@@ -150,6 +152,91 @@ def obey_config_change(log, transaction_id, config, scaling_group, state,
     return d
 
 
+def delete_group(log, trans_id, group, force):
+    """
+    Delete group based on the kind of tenant
+
+    :param log: Bound logger
+    :param str trans_id: Transaction ID of request doing this
+    :param otter.models.interface.IScalingGroup scaling_group: the scaling
+        group object
+    :param bool force: Should group be deleted even if it has servers?
+
+    :return: Deferred that fires with None
+    :raise: `GroupNotEmptyError` if group is not empty and force=False
+    """
+
+    def check_and_delete(_group, state):
+        if state.desired == 0:
+            d = trigger_convergence_deletion(log, group)
+            return d.addCallback(lambda _: state)
+        else:
+            raise GroupNotEmptyError(group.tenant_id, group.uuid)
+
+    if tenant_is_enabled(group.tenant_id, config_value):
+        return (trigger_convergence_deletion(log, group)
+                if force else group.modify_state(check_and_delete))
```

Comment:
I don't understand why this uses `modify_state`. It doesn't look like the state object is being mutated anywhere, so it seems like a no-op, OR perhaps just a roundabout way to lock the group. But since you're not locking the group for the `force` case, I'm not sure why you'd do it in the non-`force` case.